frame/election-provider-support/src/lib.rs

```diff
@@ -472,3 +477,24 @@ impl<
 		sp_npos_elections::phragmms(winners, targets, voters, Balancing::get())
 	}
 }
+
+/// A voter, at the level of abstraction of this crate.
+pub type Voter<AccountId, Bound> = (AccountId, VoteWeight, BoundedVec<AccountId, Bound>);
+
+/// Same as [`Voter`], but parameterized by an [`ElectionDataProvider`].
+pub type VoterOf<D> =
+	Voter<<D as ElectionDataProvider>::AccountId, <D as ElectionDataProvider>::MaxVotesPerVoter>;
+
+/// Extension trait to easily convert from bounded voters to unbounded ones.
+///
+/// Undesirably, this cannot be done (in safe rust) without re-allocating the entire vector, so
+/// don't use it recklessly.
```

Comment:
couldnt we return the `VoterOf` type, then we would only need to pass the inner of self, which shouldn't need realloc

Reply:
Returning `VoterOf` would be the same, it is just a type alias. The unfortunate fact is that we do `.collect`, and that's always a new allocation afaik.

Reply:
BoundedVec has inside of it just a `Vec`, so calling `bounded_vec.inner()` should give you the unbounded vec without needing to allocate anything else i think. Specifically because you need to restructure the type, I think you are doing a new allocation, which I am saying you dont need to do

Reply:
Indeed if you have just an API like: `fn foo(x: Vec<u32>)`, it is easy to pass a bounded vec to it. But if you have `fn foo(x: Vec<Vec<u32>>)`, it becomes impossible to pass `Vec<BoundedVec<u32, _>>` to it without allocation, as far as I knew before. This is because in a naive solution, you need to collect a `Vec<Vec<_>>` which is a different type from the perspective of rustc than `Vec<BoundedVec<u32, _>>`, and in principle it could have a different size, so it will allocate it again. 

There are `unsafe` ways to tell rustc "hey, `Vec` and `BoundedVec` ALWAYS have the same size, just chill", but I didn't explore it further.  

Some links about this: 
- https://stackoverflow.com/questions/45753923/can-a-vector-be-moved-and-modified-without-an-extra-allocation
-  https://stackoverflow.com/questions/63008205/how-to-access-a-vector-of-newtypes-as-the-underlying-type
-  https://stackoverflow.com/questions/48308759/how-do-i-convert-a-vect-to-a-vecu-without-copying-the-vector

In general, what I have learned  is that  nested  bounded vectors can be a PITA. 

Finally, thanks to your feedback, I came up with https://github.com/paritytech/substrate/pull/10601/commits/587f7154ec43195ccd3c1bd08664de59fbb85b03, which should for the most part, resolve all of this.